README.md
# What is flink?
flink enables any processors to communicate with an external FPGA over a serial interface. More information about the project can be found in [flink](http://flink-project.ch/start).

## flinkLinux
flink Kernel Modules for Linux offer drivers capabilities to communicate with various hardware interfaces. Among them are PCI and SPI. 
This documentation is intended for developers to understand the internals of the core driver. This facilitates the creation of new modules for other hardware interfaces. The user documentation for flinkLinux can be found in [flink](http://flink-project.ch/flink_linux).

## flink_axi
The bus module obtains its configuration data from the device tree. Alternatively, you can use hard-coded parameters by uncommenting the line "//#define CONFIG_SETTINGS_HARD_CODED" and adjusting the defines below. The device tree node should follow this structure:

```device-tree
flink_axi_0: flink_axi_registers@7aa00000 {
    compatible = "ost,flink-axi-1.0";   // do not change
    interrupt-parent = <&axi_intc_0>;   // Set this to the name of the IRQ controller used in the FPGA.
    interrupts = <0 0>;                 // do not change
    reg = <0x7aa00000 0xa000>;          // <[Base address] [length]>
    ost,flink-signal-offset = <34>;     // Signal offset to first signal nr
    ost,flink-nof-irq = <30>;           // Number of irq's
};
```
This node is tested with kernel 5.15.19-rt29-xilinx-v2022.1 --> Kernelversion 5.15.19 with the realtime patch.
Flink uses a lot of signals. Be careful with other kernels. It uses signals from (SIGRTMIN +2) up to SIGRTMAX. SIGRTMIN and (SIGRTMIN +1) has problems and should not be used!!!

## Documentation
- [Overview](doc/overview.md)
- [Bus Communication Modules](doc/bcm.md)
- [API](http://api.flink-project.ch/doc/flinklinux/html)


CHANGELOG.md
# Change Log

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## Unreleased
(v1.1.0 targeted for 2024-06-30) ([Github compare v1.0.0...master](https://github.com/flink-project/flinklinux/compare/v1.0.0...master))

### Added Features



## v1.0.0
(2023-12-13)

### Notes
This is the first official release.



flink.h
/** @file flink.h
 *  @brief Function prototypes and data structures for core module.
 *
 *  @author Martin Züger
 *  @author Urs Graf
 *  @author Patrick Good
 * 
 *  Changelog
 *  Date      Who   What
 *  28.10.23  Good  Added struct flink_device with struct flink_irq_data and struct flink_signal_data
 */
#ifndef FLINK_H_
#define FLINK_H_

#include <linux/types.h>
#include <linux/spinlock_types.h>
#include <linux/mutex.h>
#include <linux/list.h>
#include <linux/fs.h>
#include "flinkinterface/ioctl/flink_ioctl.h"

// ################# Debugging #################
// Uncomment it to see debugging information in the kernel log.
//#define DBG

// To also enable debugging in the IRQ service routine, uncomment define DBG_IRQ.
// This is disabled by default, even if debugging is enabled, because the code in the ISR is very time-critical.
#ifdef DBG
	//#define DBG_IRQ
#endif

// ############ flink error numbers ############
#define UNKOWN_ERROR -1

// ######### For compiler optimisations #########
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)

// FPGA module interface types
extern const char* fmit_lkm_lut[];

// ############ Forward declarations ############
struct flink_device;

// ############ flink private data ############
/** @brief Private data structure which is associated with a file.
 * The user library communicates with the kernel modules through read and write system calls
 * As a parameter a pointer to a file descriptor is passed. Within this structure the field @private_data
 * will hold the information about which device and subdevice will be targeted.
 */
struct flink_private_data {
	struct flink_device*    fdev;
	struct flink_subdevice* current_subdevice;
};

// ############ flink bus operations ############
/// @brief Functions to communicate with various bus communication modules
struct flink_bus_ops {
	u8  (*read8)(struct flink_device*, u32 addr);			/// read 1 byte
	u16 (*read16)(struct flink_device*, u32 addr);			/// read 2 bytes
	u32 (*read32)(struct flink_device*, u32 addr);			/// read 4 bytes
	int (*write8)(struct flink_device*, u32 addr, u8 val);		/// write 1 byte
	int (*write16)(struct flink_device*, u32 addr, u16 val);	/// write 2 bytes
	int (*write32)(struct flink_device*, u32 addr, u32 val);	/// write 4 bytes
	u32 (*address_space_size)(struct flink_device*);		/// get address space size
};

// ############ flink subdevice ############
#define MAX_NOF_SUBDEVICES 256
/// @brief Describes a subdevice
struct flink_subdevice {
	struct list_head     list;				/// Linked list of all subdevices of a device
	struct flink_device* parent;			/// Pointer to device which this subdevice belongs
	u8                   id;				/// Identifies a subdevice within a device
	u16                  function_id;		/// Identifies the function of the subdevice
	u8                   sub_function_id;	/// Identifies the subtype of the subdevice
	u8                   function_version;	/// Version of the function
	u32                  base_addr;			/// Base address (logical)
	u32                  mem_size;			/// Address space size
	u32                  nof_channels;		/// Number of channels
	u32                  unique_id;			/// unique id for this subdevice
};

// ############ flink device ############
/// @brief Describes a device
struct flink_device {
	struct list_head      list;				/// Linked list of all devices
	u8                    id;				/// Identifies a device
	u8                    nof_subdevices;	/// Number of subdevices
	struct list_head      subdevices;		/// Linked list of all subdevices of this device
	struct flink_bus_ops* bus_ops;			/// Pointer to structure defining the bus operation functions of this device
	struct module*        appropriated_module;	/// Pointer to bus interface modul used for this device 
	void*                 bus_data;			/// Bus specific data
	struct cdev*          char_device;		/// Pointer to cdev structure
	struct device*        sysfs_device;		/// Pointer to sysfs device structure
	struct list_head      hw_irq_data;		/// Linked list of requested IRQs
	u32                   nof_irqs;			/// Maximum IRQ that can be registered
	u32                   irq_offset;		/// offset for HW IRQ
	u32                   signal_offset;	/// offset for userspace signals
};

// ############ flink irq structure (two-dimensional dynamic array) ############
/// Some data is duplicated here to avoid searching during IRQ processing.
/// Be very careful if you change anything inside the code if it belongs to these structures.
/// @brief Holds all registered IRQs with the corresponding PID.
struct flink_irq_data {
	struct list_head	list;					/// List of all registered IRQs
	struct list_head	flink_process_data;		/// List to process data
	u32					irq_nr;					/// IRQ nr without offset
	u32					signal_count;			/// Registered signals (length of flink_signal_data). if(signal_count == 0) then the IRQ isn't registered
	u32					signal_nr_with_offset;	/// userspace signal nr
	u32					irq_nr_with_offset;		/// Precalculated IRQ NR to save time in IRQ routine
	spinlock_t			irq_lock;				/// Spinnlock to avoid data races between top half and ioctl call
	struct mutex		lock_for_ioctl;			/// To avoid data races when multiple processes call ioctl to add or remove an signal.

};
/// @brief This structure is used in the IRQ handler to send the appropriate signal number to the correct userspace process.
struct flink_process_data {
	struct list_head	list;		/// List of all user space processes that have requested the IRQ
	struct task_struct*	user_task;	/// User task to route IRQs per signal
};

// ############ Public functions ############
extern struct flink_device*    flink_device_alloc(void);
extern void                    flink_device_init(struct flink_device* fdev, struct flink_bus_ops* bus_ops, struct module* mod);
extern void                    flink_device_init_irq(struct flink_device* fdev, 
													 struct flink_bus_ops* bus_ops, 
													 struct module* mod, 
													 u32 nof_irq, 
													 u32 irq_offset, 
													 u32 signal_offset);
extern int                     flink_device_add(struct flink_device* fdev);
extern int                     flink_device_remove(struct flink_device* fdev);
extern int                     flink_device_delete(struct flink_device* fdev);
extern struct flink_device*    flink_get_device_by_id(u8 flink_device_id);
extern struct flink_device*    flink_get_device_by_cdev(struct cdev* char_device);
extern struct list_head*       flink_get_device_list(void);

extern struct flink_subdevice* flink_subdevice_alloc(void);
extern void                    flink_subdevice_init(struct flink_subdevice* fsubdev);
extern int                     flink_subdevice_add(struct flink_device* fdev, struct flink_subdevice* fsubdev);
extern int                     flink_subdevice_remove(struct flink_subdevice* fsubdev);
extern int                     flink_subdevice_delete(struct flink_subdevice* fsubdev);
extern struct flink_subdevice* flink_get_subdevice_by_id(struct flink_device* fdev, u8 flink_device_id);

extern struct class*           flink_get_sysfs_class(void);

extern int                     flink_select_subdevice(struct file* f, u8 subdevice, bool exclusive);

// ############ Constants ############
#define MAX_ADDRESS_SPACE 0x10000	/// Maximum address space for a flink device

// Memory addresses and offsets
#define MAIN_HEADER_SIZE		16	// byte
#define SUB_HEADER_SIZE			16	// byte
#define SUBDEV_FUNCTION_OFFSET	0x0000	// byte
#define SUBDEV_SIZE_OFFSET		0x0004	// byte
#define SUBDEV_NOFCHANNELS_OFFSET	0x0008	// byte
#define SUBDEV_UNIQUE_ID_OFFSET		0x000C	// byte
#define SUBDEV_STATUS_OFFSET		0x0010	// byte
#define SUBDEV_CONFIG_OFFSET		0x0014	// byte

// Types
#define INFO_FUNCTION_ID			0x00

// Userland types and sizes
/// @brief Structure containing information for ioctl system calls accessing single bits
struct ioctl_bit_container_t {
	uint32_t offset;	
	uint8_t  bit;		
	uint8_t  value;		
	uint8_t  subdevice;
};

// Userland types and sizes
/// @brief Structure containing information for ioctl system calls
struct ioctl_container_t {
	uint8_t  subdevice;
	uint32_t offset;
	uint8_t  size;
	void*    data;
};

// size of struct 'flink_subdevice' without linked list information (in bytes)
#define FLINKLIB_SUBDEVICE_SIZE		(sizeof(struct flink_subdevice)-offsetof(struct flink_subdevice,id))

#endif /* FLINK_H_ */


flink_pci.c
/*******************************************************************
 *   _________     _____      _____    ____  _____    ___  ____    *
 *  |_   ___  |  |_   _|     |_   _|  |_   \|_   _|  |_  ||_  _|   *
 *    | |_  \_|    | |         | |      |   \ | |      | |_/ /     *
 *    |  _|        | |   _     | |      | |\ \| |      |  __'.     *
 *   _| |_        _| |__/ |   _| |_    _| |_\   |_    _| |  \ \_   *
 *  |_____|      |________|  |_____|  |_____|\____|  |____||____|  *
 *                                                                 *
 *******************************************************************
 *                                                                 *
 *  PCI bus communication module                                   *
 *                                                                 *
 *******************************************************************/
/** @file flink_pci.c
 *  @brief PCI bus communication module. 
 * 
 *  Implements read and write functions over pci bus. 
 *
 *  @author Martin Züger
 *  @author Urs Graf
 */

#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/init.h>
#include <linux/slab.h>
#include <linux/device.h>
#include <linux/pci.h>

#include "flink.h"
#include "flink_pci.h"

//#define DBG 1
#define MODULE_NAME THIS_MODULE->name

#define BAR_0 0

MODULE_AUTHOR("Martin Zueger <martin@zueger.eu>");
MODULE_DESCRIPTION("fLink PCI/PCIe module");
MODULE_LICENSE("Dual BSD/GPL");

// ############ Module parameters ############
static unsigned short vid = 0x1172;
static unsigned short pid = 0x0004;

module_param(vid, ushort, 0444);
MODULE_PARM_DESC(vid, "PCI vendor ID, eg. '0x1172' for Altera");
module_param(pid, ushort, 0444);
MODULE_PARM_DESC(pid, "PCI product ID, eg. '0x000'");

// ############ Bus communication functions ############
u8 pci_read8(struct flink_device* fdev, u32 addr) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		return ioread8(pci_data->base_addr + addr);
	}
	return 0;
}

u16 pci_read16(struct flink_device* fdev, u32 addr) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		return ioread16(pci_data->base_addr + addr);
	}
	return 0;
}

u32 pci_read32(struct flink_device* fdev, u32 addr) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	u32* read_address = (u32*)(pci_data->base_addr + addr);
	if(pci_data != NULL) {
		#if defined(DBG)
			printk(KERN_DEBUG "[%s] Reading 32 bit from PCI device (flink device id %u) at address 0x%p...", MODULE_NAME, fdev->id, read_address);
		#endif
		return ioread32(read_address);
	}
	else {
		#if defined(DBG)
			printk(KERN_ERR "[%s] Reading 32 bit from PCI device (flink device id %u) at address 0x%p failed!", MODULE_NAME, fdev->id, read_address);
		#endif
	}
	return 0;
}

int pci_write8(struct flink_device* fdev, u32 addr, u8 val) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		iowrite8(val, pci_data->base_addr + addr);
		return 0;
	}
	return -1;
}

int pci_write16(struct flink_device* fdev, u32 addr, u16 val) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		iowrite16(val, pci_data->base_addr + addr);
		return 0;
	}
	return -1;
}

int pci_write32(struct flink_device* fdev, u32 addr, u32 val) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		iowrite32(val, pci_data->base_addr + addr);
		return 0;
	}
	return -1;
}

u32 pci_address_space_size(struct flink_device* fdev) {
	struct flink_pci_data* pci_data = (struct flink_pci_data*)fdev->bus_data;
	if(pci_data != NULL) {
		return (u32)(pci_data->mem_size - PCI_CONFIG_SIZE);
	}
	return 0;
}

struct flink_bus_ops pci_bus_ops = {
	.read8              = pci_read8,
	.read16             = pci_read16,
	.read32             = pci_read32,
	.write8             = pci_write8,
	.write16            = pci_write16,
	.write32            = pci_write32,
	.address_space_size = pci_address_space_size
};

// ############ Device handling ############
static struct flink_device* create_flink_pci_device(struct flink_bus_ops* bus_ops, struct pci_dev* pci_device, void __iomem* base_ptr, unsigned long length) {
	struct flink_pci_data* pci_data = kmalloc(sizeof(struct flink_pci_data), GFP_KERNEL);
	struct flink_device* fdev = flink_device_alloc();
	
	if(pci_data != NULL && fdev != NULL) {
		pci_data->pci_device = pci_device;
		pci_data->base_addr = base_ptr + BASE_OFFSET;
		pci_data->mem_size = length;
		
		flink_device_init(fdev, bus_ops, THIS_MODULE);
		fdev->bus_data = pci_data;
		return fdev;
	}
	return NULL;
}

// ############ Initialization and cleanup ############
static int __init flink_pci_init(void) {
	int error = 0;
	struct flink_device* flink_pci_dev;
	struct pci_dev* pci_device;
	void __iomem* mmio_base_ptr;
	unsigned long mmio_length;
	
	#if defined(DBG)
		printk(KERN_DEBUG "[%s] Initializing module with parameters 'vid=%x, pid=%x'", MODULE_NAME, vid, pid);
	#endif

	// Get PCI device struct
	pci_device = pci_get_device(vid, pid, NULL);
	if(pci_device == NULL) {
		printk(KERN_ALERT "[%s] ERROR: PCIe device not found!", MODULE_NAME);
		goto err_pci_device_not_found;
	}
	
	// Initialize and enable the PCI device
	error = pci_enable_device(pci_device);
	if(error) {
		printk(KERN_ALERT "[%s] ERROR: Unable to enable PCI device!", MODULE_NAME);
		goto err_pci_enable_device;
	}
	
	// Reserve PCI memory resources
	error = pci_request_regions(pci_device, KBUILD_MODNAME);
	if(error) {
		printk(KERN_ALERT "[%s] ERROR: Memory region request failed!", MODULE_NAME);
		goto err_pci_region_request;
	}
	
	// I/O Memory mapping
	mmio_length = pci_resource_len(pci_device, BAR_0);
	mmio_base_ptr = pci_iomap(pci_device, BAR_0, mmio_length);
	#if defined(DBG)
		printk(KERN_DEBUG "[%s] PCI resource I/O memory mapping:", MODULE_NAME);
		printk(KERN_DEBUG "  -> Base address:  0x%p", mmio_base_ptr);
		printk(KERN_DEBUG "  -> Memory length: 0x%lx (%lu bytes)", mmio_length, mmio_length);
	#endif
	if(mmio_base_ptr == NULL){
		printk(KERN_ALERT "[%s] ERROR: I/O Memory mapping failed!", MODULE_NAME);
		goto err_pci_iomap;
	}
	
	flink_pci_dev = create_flink_pci_device(&pci_bus_ops, pci_device, mmio_base_ptr, mmio_length);
	flink_device_add(flink_pci_dev);
	
	// All done
	printk(KERN_INFO "[%s] Module sucessfully loaded", MODULE_NAME);
	
	return 0;

// ---- ERROR HANDLING ----
	err_pci_iomap:
		pci_release_regions(pci_device);
	
	err_pci_region_request:
		pci_disable_device(pci_device);
	
	err_pci_enable_device:
		pci_device = NULL;
	
	err_pci_device_not_found:
		// nothing to do
	
	return error;
}

static void __exit flink_pci_exit(void) {
	struct flink_device* fdev;
	struct flink_device* fdev_next;
	struct flink_pci_data* pci_data;
	
	#if defined(DBG)
		printk(KERN_DEBUG "[%s] Looking for devices which are appropriated to this module...", MODULE_NAME);
	#endif
	
	list_for_each_entry_safe(fdev, fdev_next, flink_get_device_list(), list) {
		if(fdev->appropriated_module == THIS_MODULE) {
			#if defined(DBG)
				printk(KERN_DEBUG "[%s] Device with id '%u' is appropriated to this module and will be removed!", MODULE_NAME, fdev->id);
			#endif
			pci_data = (struct flink_pci_data*)(fdev->bus_data);
			pci_release_regions(pci_data->pci_device);
			pci_disable_device(pci_data->pci_device);
			kfree(pci_data);
			flink_device_remove(fdev);
			flink_device_delete(fdev);
		}
		else {
			#if defined(DBG)
				printk(KERN_DEBUG "[%s] Device with id '%u' is not appropriated to this module!", MODULE_NAME, fdev->id);
			#endif
		}
	}
	printk(KERN_INFO "[%s] Module sucessfully unloaded", MODULE_NAME);
}

module_init(flink_pci_init);
module_exit(flink_pci_exit);


zynq/flink_axi.c
/*******************************************************************
 *   _________     _____      _____    ____  _____    ___  ____    *
 *  |_   ___  |  |_   _|     |_   _|  |_   \|_   _|  |_  ||_  _|   *
 *    | |_  \_|    | |         | |      |   \ | |      | |_/ /     *
 *    |  _|        | |   _     | |      | |\ \| |      |  __'.     *
 *   _| |_        _| |__/ |   _| |_    _| |_\   |_    _| |  \ \_   *
 *  |_____|      |________|  |_____|  |_____|\____|  |____||____|  *
 *                                                                 *
 *******************************************************************
 *                                                                 *
 *  AXI (Advanced eXtensible Interface) communication module for   *
 *  Zync7000 SoC                                                   *
 *                                                                 *
 *******************************************************************/

/** @file flink_axi.c
 *  @brief AXI (Advanced eXtensible Interface) communication module.
 *
 *  Implements read and write functions over AXI Bus.
 *
 *  @author Patrick Good
 */


#include <linux/module.h>
#include <linux/platform_device.h>
#include <linux/io.h>
#include <linux/of_device.h>
#include <linux/of_address.h>
#include <linux/slab.h>
#include <linux/of_irq.h>
#include <linux/of_address.h>

#include "../flink.h"

/* Config option (device tree or hard coded) If you want to configure 
 * this module with hard coded values, just uncomment the define below 
 * and set your values to the defines. Otherwise it will look for a 
 * compatible node in the device tree.
**/
//#define CONFIG_SETTINGS_HARD_CODED
#ifdef CONFIG_SETTINGS_HARD_CODED
	#define AXI_BASE_ADDR 		0x7aa00000 // if you adust it here also adjust the number in define NODE_NAME
	#define AXI_Range_LENGHT 	0x9000
	#define IRQ_OFFSET 			55
	#define SIGNAL_OFFSET 		34
	#define NOF_IRQs			30
	#define NODE_NAME           "flink_axi@7aa00000"
#else
	#define COMPATIBLE_NODE "ost,flink-axi-1.0"
#endif

// ############ Module infos ############
MODULE_AUTHOR("Patrick Good");
MODULE_DESCRIPTION("Flink AXI communication Module");
MODULE_LICENSE("Dual BSD/GPL");

// ############ Module Defines ############
#define MOD_VERSION "0.1.0"
#define MODULE_NAME THIS_MODULE->name

#define PLATFORM_DEV_NAME "flink_axi_driver"
#define PLATFORM_DEV_ID -1

// ############ Function Prototypes ############
static u8 flink_axi_read8(struct flink_device* fdev, u32 addr);
static u16 flink_axi_read16(struct flink_device* fdev, u32 addr);
static u32 flink_axi_read32(struct flink_device* fdev, u32 addr);
static int flink_axi_write8(struct flink_device* fdev, u32 addr, u8 val);
static int flink_axi_write16(struct flink_device* fdev, u32 addr, u16 val);
static int flink_axi_write32(struct flink_device* fdev, u32 addr, u32 val);
static u32 flink_axi_address_space_size(struct flink_device* fdev);

static int flink_axi_probe(struct platform_device *pdev);
static int flink_axi_remove(struct platform_device *pdev);

// ############ Module Data Structs ############
#ifdef CONFIG_SETTINGS_HARD_CODED
	static struct platform_driver flink_axi_driver = {
	 	.driver = {
			.name = PLATFORM_DEV_NAME,
			.owner = THIS_MODULE,
	    },
	    .probe = flink_axi_probe,
	    .remove = flink_axi_remove,
	};

	static struct platform_device flink_axi_device = {
    	.name = PLATFORM_DEV_NAME,
    	.id = PLATFORM_DEV_ID,
	};
#else
	static const struct of_device_id my_driver_of_match[] = {
	    { .compatible = COMPATIBLE_NODE},
	    {},
	};

	static struct platform_driver flink_axi_driver = {
	 	.driver = {
			.name = PLATFORM_DEV_NAME,
			.owner = THIS_MODULE,
			.of_match_table = my_driver_of_match,
	    },
	    .probe = flink_axi_probe,
	    .remove = flink_axi_remove,
	};
#endif

struct flink_axi_bus_data
{
	void __iomem *base;
	resource_size_t hardwareAddressBase;
	resource_size_t size;
};

struct flink_bus_ops flink_axi_bus_ops =
{
	.read8              = flink_axi_read8,
	.read16             = flink_axi_read16,
	.read32             = flink_axi_read32,
	.write8             = flink_axi_write8,
	.write16            = flink_axi_write16,
	.write32            = flink_axi_write32,
	.address_space_size = flink_axi_address_space_size
};

// ############ Module Bus Operations ############
static inline u8 flink_check_offset(struct flink_axi_bus_data* d, u32 offset) {
	return d->size > offset;
}

static u8 flink_axi_read8(struct flink_device* fdev, u32 addr) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		return ioread8(d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the ioread8 operation\n", MODULE_NAME);
	}
	return 0;
}

static u16 flink_axi_read16(struct flink_device* fdev, u32 addr) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		return ioread16(d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the ioread16 operation\n", MODULE_NAME);
	}
	return 0;
}

static u32 flink_axi_read32(struct flink_device* fdev, u32 addr) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		return ioread32(d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the ioread32 operation\n", MODULE_NAME);
	}
	return 0;
}

static int flink_axi_write8(struct flink_device* fdev, u32 addr, u8 val) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		iowrite8(val, d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the iowrite8 operation\n", MODULE_NAME);
	}
	return 0;
}

static int flink_axi_write16(struct flink_device* fdev, u32 addr, u16 val) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		iowrite16(val, d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the iowrite16 operation\n", MODULE_NAME);
	}
	return 0;
}

static int flink_axi_write32(struct flink_device* fdev, u32 addr, u32 val) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	if (likely(d != NULL && flink_check_offset(d,addr))) {
		iowrite32(val, d->base + addr);
	} else {
		printk(KERN_ERR "[%s] Failed to perform the iowrite32 operation\n", MODULE_NAME);
	}
	return 0;
}

static u32 flink_axi_address_space_size(struct flink_device* fdev) {
    struct flink_axi_bus_data* d = (struct flink_axi_bus_data*)fdev->bus_data;
	return (u32)(d->size);
}

// ############ Platform Driver Probe And Remove ############
static int flink_axi_probe(struct platform_device *pdev)
{
	struct device_node *np = pdev->dev.of_node;
    u32 irq_offset = 0;
	u32 signal_offset = 0;
	u32 reg[2];
	u32 nof_irq = 0;
    int ret = 0;
//    void __iomem *base;
    int err = 0;
    struct flink_device *fdev;
	struct flink_axi_bus_data *bus_data;

	bus_data = kzalloc(sizeof(struct flink_axi_bus_data), GFP_KERNEL);
	if (unlikely(!bus_data)) {
		printk(KERN_ERR "[%s] Failed to allocate mem for bus data\n", MODULE_NAME);
		ret = -ENOMEM;
		goto bus_data_alloc_failure;
	}

	#ifdef CONFIG_SETTINGS_HARD_CODED
		bus_data->hardwareAddressBase = AXI_BASE_ADDR;
		bus_data->size = AXI_Range_LENGHT;
		irq_offset = IRQ_OFFSET;
		signal_offset = SIGNAL_OFFSET;
		nof_irq = NOF_IRQs;

		if (unlikely(!request_mem_region(bus_data->hardwareAddressBase, bus_data->size, NODE_NAME))) {
			printk(KERN_ERR "[%s] Failed to request AXI memory region", MODULE_NAME);
			ret = -ENOMEM;
			goto mem_request_failure;
		}

    	bus_data->base = ioremap(bus_data->hardwareAddressBase, bus_data->size);
		if (unlikely(!(bus_data->base))) {
        	printk(KERN_ERR "[%s] Failed to map AXI memory\n", MODULE_NAME);
        	ret = -ENOMEM;
			goto mem_iomap_failure;
		}
	#else
		err = of_property_read_u32(np, "ost,flink-nof-irq", &nof_irq);
    	if (unlikely(err < 0)) {
        	printk(KERN_ERR "[%s] Failed to read 'ost,flink-nof-irq' property. Error Nr: %d\n", MODULE_NAME, err);
			ret = -ENOMEM;
        	goto read_poperties_failure;
    	}
		err = of_property_read_u32(np, "ost,flink-signal-offset", &signal_offset);
    	if (unlikely(err < 0)) {
        	printk(KERN_ERR "[%s] Failed to read 'ost,flink-signal-offset' property. Error Nr: %d\n", MODULE_NAME, err);
			ret = -ENOMEM;
        	goto read_poperties_failure;
    	}
		err = of_property_read_u32_array(np, "reg", reg, 2);
    	if (unlikely(err < 0)) {
        	printk(KERN_ERR "[%s] Failed to read hardware address register (reg) property. Error Nr: %d\n", MODULE_NAME, err);
			ret = -ENOMEM;
        	goto read_poperties_failure;
    	}
		bus_data->hardwareAddressBase = reg[0];
		bus_data->size = reg[1];
		irq_offset = of_irq_get(np, 0);
		if (unlikely(irq_offset < 0)) {
			printk(KERN_ERR "[%s] Failed to get the base irq number from IRQ controller. Error Nr: %d\n", MODULE_NAME, err);
			ret = -ENOMEM;
        	goto read_poperties_failure;
		}
		if (unlikely(!request_mem_region(bus_data->hardwareAddressBase, bus_data->size, MODULE_NAME))) {
			printk(KERN_ERR "[%s] Failed to request AXI memory region", MODULE_NAME);
			ret = -ENOMEM;
			goto mem_request_failure;
		}
		bus_data->base = of_iomap(np, 0);
    	if (unlikely(!(bus_data->base))) {
			printk(KERN_ERR "[%s] Failed to map AXI memory\n", MODULE_NAME);
        	ret = -ENOMEM;
			goto mem_iomap_failure;
		}
	#endif

	#ifdef DBG
		#ifdef CONFIG_SETTINGS_HARD_CODED
			printk(KERN_DEBUG "[%s] Hard coded values are:", MODULE_NAME);
		#else
			printk(KERN_DEBUG "[%s] Values from Device tree are:", MODULE_NAME);
		#endif
		printk(KERN_DEBUG "  --> HW address:     %#lx\n", bus_data->hardwareAddressBase);
		printk(KERN_DEBUG "  --> HW vir address: %#lx\n", bus_data->base);
		printk(KERN_DEBUG "  --> HW size:        %#lx\n", bus_data->size);
		printk(KERN_DEBUG "  --> IRQ offset:     %lu\n", irq_offset);
		printk(KERN_DEBUG "  --> Siagnal offset: %lu\n", signal_offset);
		printk(KERN_DEBUG "  --> nof IRQ's:      %lu\n", nof_irq);
	#endif
	
	fdev = flink_device_alloc();
	if (unlikely(!fdev)) {
		printk(KERN_ERR "[%s] Failed to allocate flink device\n", MODULE_NAME);
		ret = -ENOMEM;
		goto fdev_alloc_failure;
	}

    // setup flink device
	flink_device_init_irq(fdev, &flink_axi_bus_ops, THIS_MODULE, nof_irq, irq_offset, signal_offset);
	fdev->bus_data = bus_data;
	#ifdef DBG
		printk(KERN_DEBUG "[%s] Create flink device...", MODULE_NAME);
	#endif
	if(unlikely(flink_device_add(fdev) < 0)){
		printk(KERN_ERR "[%s] Faild to add flink device", MODULE_NAME);
		ret = -ENOMEM;
		goto flink_add_failure;
	}
	printk(KERN_INFO "[%s] Flink device created", MODULE_NAME);
    return 0;


	flink_add_failure:
		flink_device_delete(fdev);
	fdev_alloc_failure:
		iounmap(bus_data->base);
	mem_iomap_failure:
		release_mem_region(bus_data->hardwareAddressBase, bus_data->size);
	mem_request_failure:
	read_poperties_failure:
		kfree(bus_data);
	bus_data_alloc_failure:
		// nothing to clean up
		printk(KERN_ERR "[%s] Failed to initialise flink-AXI driver\n", MODULE_NAME);
	return ret;
}

static int flink_axi_remove(struct platform_device *pdev)
{
    struct flink_device *fdev;
	struct flink_device *fdev_next;
	struct flink_axi_bus_data *bus_data;

	#ifdef DBG
		printk(KERN_DEBUG "[%s] AXI platform device removing", MODULE_NAME);
	#endif
	list_for_each_entry_safe(fdev, fdev_next, flink_get_device_list(), list) {
		if(fdev->appropriated_module == THIS_MODULE) {
			bus_data = (struct flink_axi_bus_data *)(fdev->bus_data);
			flink_device_remove(fdev);
			flink_device_delete(fdev);
			iounmap(bus_data->base);
			release_mem_region(bus_data->hardwareAddressBase, bus_data->size);
			kfree(bus_data);
		}
	}
	return 0;
}

// ############ Module Initialization and cleanup ############
static int __init axi_init(void)
{
	int err = 0;

	#ifdef CONFIG_SETTINGS_HARD_CODED	
    	err = platform_device_register(&flink_axi_device);
		if (err < 0) {
			printk(KERN_ERR "[%s] cannot register device: %d\n", MODULE_NAME, err);
			goto exit;
		}

		err = platform_driver_register(&flink_axi_driver);
		if (err < 0) {
			printk(KERN_ERR "[%s] unable to register driver: %d\n", MODULE_NAME, err);
			goto exit_unregister_device;
		}
	#else
		err = platform_driver_register(&flink_axi_driver);
		if (err < 0) {
			printk(KERN_ERR "[%s] unable to register driver: %d\n", MODULE_NAME, err);
			goto exit;
		}
	#endif

	#ifdef DBG
		printk(KERN_DEBUG "[%s] Module successfully loaded\n", MODULE_NAME);
	#endif
	return 0;

	#ifdef CONFIG_SETTINGS_HARD_CODED
		exit_unregister_device:
			platform_device_unregister(&flink_axi_device);
	#endif
	exit:
		printk(KERN_ERR "[%s] Init failed\n", MODULE_NAME);
	return err;
}

static void __exit axi_exit(void)
{
    platform_driver_unregister(&flink_axi_driver);
	#ifdef CONFIG_SETTINGS_HARD_CODED	
    	platform_device_unregister(&flink_axi_device);
	#endif
	#ifdef DBG
		printk(KERN_DEBUG "[%s] Module successfully unloaded\n", MODULE_NAME);
	#endif
}

module_init(axi_init);
module_exit(axi_exit);
